### ISSUE TITLE
Issues now that terragrunt-action "updates permissions" of workspace files

### ISSUE THREAD

**Issue by MidwestMetropolis** [gruntwork-io/terragrunt-action#60]
Hello,

I wanted to report that I am not able to use the latest version of terragrunt-action, either because my runner instance has AWS Systems Manager Session Manager enabled, or because I am using a self-hosted runner. I have permissions issues with post-job checkout file/etc. cleanup. I think there is some issue between the ubuntu/ssm-user/runner users (i.e., the "runner" user in the terragrunt container that you publish from this repo). After terragrunt-action does its permissions updating, my checked-out repo files on the runner are owned by ssm-user (id 1001) (the runner user in the terragrunt container is prob. also id 1001?).

I wanted to recommend rethinking the current approach of having a runner user within the terragrunt container, largely because the GitHub Hosted Runners [documentation](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#docker-container-filesystem) says:

_/github/workspace - Note: GitHub Actions **must be run by the default Docker user (root)**. Ensure your Dockerfile does not set the USER instruction, otherwise you will not be able to access GITHUB_WORKSPACE._

Your terragrunt Dockerfile does use the USER instruction (setting the user to "runner"). Perhaps there is another approach? Thanks for your work!

**Comment by arsci** [comment]
Thanks for reporting this! I'm looking into the issue and will report back when I have more info, or a fix ready.

**Comment by arsci** [comment]
@MidwestMetropolis could you share some additional details on your self-hosted runner? I think I have a fix just about ready, but I'd like to test as close to your use case as possible, and ideally add an automated test to cover it so we don't unknowingly break it in future updates

**Comment by MidwestMetropolis** [comment]
I can't share a ton of the code as it's for work, but I can give some more information.

My customized Ubuntu base image is derived from (I'm using Vagrant):
`"ami-053b0d53c279acc90" # Canonical, Ubuntu, 22.04 LTS, amd64 jammy image build on 2023-05-16`

I install AWS CLI on it (per Amazon's AWS CLI instructions) with:
```
mkdir ~/temp  #this folder gets used later in my scripts

curl -o ~/temp/awscliv2.zip -fsSL https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip
unzip ~/temp/awscliv2.zip -d ~/temp
sudo ~/temp/aws/install
```

I install tfenv/Terragrunt like so:
```
# tfenv
git clone --depth=1 https://github.com/tfutils/tfenv.git ~/.tfenv
sudo ln -s ~/.tfenv/bin/* /usr/local/bin

# Terragrunt
curl -o ~/temp/terragrunt -fsSL https://github.com/gruntwork-io/terragrunt/releases/download/v0.57.6/terragrunt_linux_amd64
chmod +x ~/temp/terragrunt
sudo mv ~/temp/terragrunt /usr/local/bin
sudo chown $(stat -c "%U:%G" /usr/local/bin) /usr/local/bin/terragrunt
```

I install Docker on it (per the instructions on their site) with:
```
sudo apt install ca-certificates curl
sudo install -m 0755 -d /etc/apt/keyrings
sudo curl -o /etc/apt/keyrings/docker.asc -fsSL https://download.docker.com/linux/ubuntu/gpg
sudo chmod a+r /etc/apt/keyrings/docker.asc
echo \
  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \
  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
  sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
sudo apt update
sudo apt install -y docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
sudo usermod -aG docker $USER
```

Then I have a self-hosted runner image, derived from the image above, that gets set up as a GitHub runner via:
```
# from GitHub repo settings ("Actions" -> "Runners" -> "New self-hosted runner")
mkdir actions-runner && cd actions-runner
curl -o actions-runner-linux-x64-2.315.0.tar.gz -fsSL https://github.com/actions/runner/releases/download/v2.315.0/actions-runner-linux-x64-2.315.0.tar.gz
echo "6362646b67613c6981db76f4d25e68e463a9af2cc8d16e31bfeabe39153606a0  actions-runner-linux-x64-2.315.0.tar.gz" | shasum -a 256 -c  # the double space is required for shasum
tar xzf ./actions-runner-linux-x64-2.315.0.tar.gz
```
Note that you still have to "activate" the runner manually via the instructions on GitHub's site when you add a "New self-hosted runner".

The GitHub Actions workflow section that I used to run terragrunt-action on my self-hosted runner looked like:
```
- name: Tear down Amazon RDS instance (Terragrunt destroy)
  uses: gruntwork-io/terragrunt-action@v2.0.2
  env:
    AWS_PROFILE: devops-secret-project-name-dev-terraform
    AWS_REGION: us-east-1
    TERRAGRUNT_RDS_WORKING_FOLDER: secret-project-name/terragrunt/secret-project-name/aws/devops-secret-project-name-dev/us-east-1/rds
    INPUT_PRE_EXEC_1: |
      # There was some AWS credentials setup here
  with:
    tf_version: 1.7.5
    tg_version: 0.55.21
    tg_dir: ${{ env.TERRAGRUNT_RDS_WORKING_FOLDER }}
    tg_command: plan  # Eventually this would have been a destroy command. When
                      # I saw the permissions changes/issues after running terragrunt-action
                      # I stopped using it in my workflow. As I mentioned, I saw some
                      # permissions issues related to the Git cleanup steps that happen late in the
                      # workflow. Also, I saw that some files, after running terragrunt-action, were
                      # newly owned by user "ssm-user" (id 1001) after the action completed id. That is
                      # the user created for AWS Systems Manager Session Manager by the SSM Agent.
                      # Note that in order for instances to use this features, they have to have
                      # an IAM role with the AWS-managed policy "AmazonSSMManagedInstanceCore".
```

Depending on how you tyr to address this issue, most of the above may simply be TMI. E.g., if you just use root in your terragrunt container vs. some other trick (like utilizing a higher-ID user, and continuing to try to change/undo permissions over the course of the action's execution). I'm largely just guessing, as I only understand some of the details.

Thank you!

**Comment by arsci** [comment]
Thanks @MidwestMetropolis! I appreciate the detail. I've got a fix just about ready, we are just running tests and reviewing. I'll let you know when it's ready and released, and then I can follow up after you've had a chance to test it.

**Comment by arsci** [comment]
@MidwestMetropolis we just released [v2.1.1](https://github.com/gruntwork-io/terragrunt-action/releases/tag/v2.1.1) that should address the permissions problems. This issue auto-closed, but I re-opened it until we confirm you are able to use the latest version.

**Comment by arsci** [comment]
I'm going to close this issue, please feel free to re-open the issue if the latest releases have not resolved the problem.
